addons/stock/models/stock_rule.py

```diff
@@ -35,36 +35,143 @@ class ProcurementRule(models.Model):
+    def _onchange_route(self):
+        """ If a route is set and it exists a route company, ensure that the
+        rule's company is the same than the route's company.
+        """
+        company_id = self.route_id.company_id
+        if company_id:
+            self.company_id = company_id
+            if self.picking_type_id.warehouse_id.company_id != company_id:
+                self.picking_type_id = False
+        domain = {'company_id': company_id and [('id', '=', company_id.id)] or []}
+        return {'domain': domain}
+
+    def _get_message_values(self):
+        """ Return the source, destination and picking_type apply on a stock
+        rule. The purpose of this function is to avoid code duplication in
+        _get_message_dict functions since it often require those data.
+        """
+        source = self.location_src_id and self.location_src_id.display_name or _('Source Location')
+        destination = self.location_id and self.location_id.display_name or _('Destination Location')
+        operation = self.picking_type_id and self.picking_type_id.name or _('Operation Type')
+        return source, destination, operation
+
+    def _get_message_dict(self):
+        """ Return a dict with the different possible message used for the
+        rule message. It should return one message for each stock.rule action
+        (except push and pull). This function is override in mrp and
+        purchase_stock in order to complete the dictionary.
+        """
+        message_dict = {}
+        source, destination, operation = self._get_message_values()
+        if self.action in ('push', 'pull', 'pull_push'):
+            suffix = ""
+            if self.procure_method == 'make_to_order' and self.location_src_id:
+                suffix = _("<br>A need is created in <b>%s</b> and a rule will be triggered to fulfill it.") % (source)
+            message_dict = {
+                'pull': _('When products are needed in <b>%s</b>, <br/> <b>%s</b> are created from <b>%s</b> to fulfill the need.') % (destination, operation, source) + suffix,
+                'push': _('When products arrive in <b>%s</b>, <br/> <b>%s</b> are created to send them in <b>%s</b>.') % (source, operation, destination)
+            }
+        return message_dict
+
+    @api.depends('action', 'location_id', 'location_src_id', 'picking_type_id', 'procure_method')
+    def _compute_action_message(self):
+        """ Generate dynamicaly a message that describe the rule purpose to the
+        end user.
+        """
+        for rule in self.filtered(lambda rule: rule.action):
+            message_dict = rule._get_message_dict()
+            message = message_dict.get(rule.action) and message_dict[rule.action] or ""
+            if rule.action == 'pull_push':
+                message = message_dict['pull'] + "<br/><br/>" + message_dict['push']
+            rule.rule_message = message
+
+    def _apply(self, move):
+        """ Apply a rule on a move.
```

Comment:
apply a push rule

Reply:
shoudln't we prefix all methods related to push rule with _push, same with _pull?

Reply:
_run_push